src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -358,12 +358,162 @@ internal void WriteEntry(EntityDescriptor entityDescriptor, IEnumerable<LinkDesc
+                //first item in determining the type of the descriptor.
+                Descriptor item = relatedLinks.Value.FirstOrDefault();
+
+                if (item is EntityDescriptor entDescriptor)
+                {
+                    relatedEnd = new LinkDescriptor(entDescriptor.ParentEntity, entDescriptor.ParentProperty, entDescriptor.Entity, this.requestInfo.Model);
+                }
+                else
+                {
+                    relatedEnd = item as LinkDescriptor;
+                }
+
+                navigationLink.Name = relatedEnd.SourceProperty;
+                bool isCollection = clientType.GetProperty(relatedEnd.SourceProperty, UndeclaredPropertyBehavior.ThrowException).IsEntityCollection;
+
+                odataWriter.WriteNestedResourceInfoStart(navigationLink);
+
+                odataWriter.WriteNestedResourceInfoStart(navigationLink, relatedEnd.Source, relatedEnd.Target);
+
+                if (isCollection)
+                {
+                    WriteResourceSet(relatedLinks.Value, linkDescriptors, bulkUpdateGraph, odataWriter);
+                }
+                else
+                {
+                    //we can have a related single entry ~ this will automatically be a replace operation.
+                    if (item.DescriptorKind == DescriptorKind.Entity)
+                    {
+                        EntityDescriptor relatedEntityDescriptor = item as EntityDescriptor;
+                        WriteDeepInsertEntry(relatedEntityDescriptor, linkDescriptors, bulkUpdateGraph, odataWriter);
+                    }
+                    else if (item.DescriptorKind == DescriptorKind.Link)
+                    {
+                        odataWriter.WriteEntityReferenceLink(new ODataEntityReferenceLink() { Url = navigationLink.Url }, relatedEnd.Source, relatedEnd.Target);
+                    }
+                }
+
+                odataWriter.WriteNestedResourceInfoEnd(navigationLink, relatedEnd.Source, relatedEnd.Target);
+                odataWriter.WriteNestedResourceInfoEnd();
+            }
+        }
+
         /// <summary>
         /// Writes the feed entry
         /// </summary>
-        /// <param name="descriptors">A list of descriptors to be written</param>
-        /// <param name="linkDescriptors">Descriptor links for some descriptors.</param>
-        /// /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/></param>
+        /// <param name="descriptors">A list of descriptors to be written.</param>
+        /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/>.</param>
+        /// <param name="oDataWriter">An instance of the <see cref="ODataWriterWrapper"/>.</param>
+        internal void WriteResourceSet(IReadOnlyList<Descriptor> descriptors, Dictionary<Descriptor, List<LinkDescriptor>> linkDescriptors, BulkUpdateGraph bulkUpdateGraph, ODataWriterWrapper oDataWriter)

```

Comment:
You are missing a doc comment for `linkDescriptors` param.

Reply:
refactored to remove `linkDescriptors` parameter